pkg/crosscluster/logical/batch_handler_test.go

```diff
@@ -0,0 +1,391 @@
+		map[descpb.ID]sqlProcessorTableConfig{
+			desc.GetID(): {
+				srcDesc: desc,
+			},
+		},
+	)
+	require.NoError(t, err)
+	return handler, desc
+}
+
+// addAndRemoveColumn is a metamorphic test option that randomly decides whether
+// to add and remove a column to trigger a rebuild of the primary key.
+var addAndRemoveColumn = metamorphic.ConstantWithTestBool("batch-handler-exhaustive-rebuild-primary-index", false)
+
+// uniqueConstraint is a metamorphic test option that randomly decides whether
+// to add a unique constraint to the table.
+var uniqueConstraint = metamorphic.ConstantWithTestBool("batch-handler-exhaustive-unique-constraint", false)
+
+func TestBatchHandlerExhaustive(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	// This test is an "exhaustive" test of the batch handler. It tries to test
+	// cross product of every possible (replication event type, local value,
+	// previous value, lww win).
+	//
+	// This doesn't test
+	// 1. Batches containing multiple rows or multiple events for the same row.
+	// 2. Events that are supposed to be DLQ'd, like unique constraint
+	// violations or foreign key violations.
+
+	ctx := context.Background()
+	srv, sqlDB, _ := serverutils.StartServer(t, base.TestServerArgs{})
+	defer srv.Stopper().Stop(ctx)
+	s := srv.ApplicationLayer()
+	runner := sqlutils.MakeSQLRunner(sqlDB)
+
+	// TODO(jeffswenson): improve this test so that it uses a random schema.
+
+	// Create a test table with an index
+	runner.Exec(t, `
+		CREATE TABLE test_table (
+			id INT PRIMARY KEY,
+			value STRING
+		)
+	`)
+
+	if uniqueConstraint {
+		runner.Exec(t, `ALTER TABLE test_table ADD CONSTRAINT unique_value UNIQUE (value)`)
+	}
+	// TODO(jeffswenson): enable this metamorphic option after fixing #146465.
+	if false && addAndRemoveColumn {
+		runner.Exec(t, `ALTER TABLE test_table ADD COLUMN temp_col INT`)
+		runner.Exec(t, `ALTER TABLE test_table DROP COLUMN temp_col`)
+	}
+
+	handler, desc := newBatchHandler(t, s, "test_table")
+	defer handler.ReleaseLeases(ctx)
+
+	// TODO(jeffswenson): test the other handler types.
+	eventBuilder := newKvEventBuilder(t, desc.TableDesc())
+
+	type previousValue string
+	const (
+		previousValuePresent previousValue = "present"
+		previousValueAbsent  previousValue = "absent"
+	)
+	previousValues := []previousValue{
+		previousValuePresent,
+		previousValueAbsent,
+	}
+
+	type localValue string
+	const (
+		// The local value is the same as the previous value supplied by the
+		// event.
+		localValueMatch localValue = "previous-value"
+		// The local value is different from the previous value supplied by the
+		// event.
+		localValueMismatch localValue = "different-value"
+		// The local value never existed in the database.
+		localValueNull localValue = "null-value"
+		// The local value is a tombstone.
+		localValueTombstone localValue = "tombstone-value"
+		// The local value is the same as the previous value supplied by the
```

Comment:
nit: docstring copied over from localValueMatch.

Reply:
i believe the main difference between the two is that for localValueReplay, the local value equals the value about to be inserted, whereas for localValuesMatch they do not.